### ISSUE TITLE
Applying MAPIE in 'prefit' mode to catboost model exhaust all RAM 

### ISSUE THREAD

**Issue by artem-math** [scikit-learn-contrib/MAPIE#328]
**Describe the bug**
Applying MAPIE(v0.6, 0.6.5) in 'prefit' mode to a catboost (v1.1, v1.2) model exhaust all RAM 

**To Reproduce**
1. Go to "..."
2. Click on "..."
3. Scroll down to "..."

**Expected behavior**
....fit a catboost model...
mapie = MapieRegressor(estimator=model, cv="prefit", n_jobs=1)
mapie = mapie.fit(X_calib, y_calib)
### <-- Until this moment everything is fine

y_pred, y_pis = mapie.predict(X_calib, alpha=0.5) ## Here the RAM goes up from 4GB to 12GB (max) independently of how elaborate is the catboost model.

**Comment by thibaultcordier** [comment]
Hello @artem-math, thank you for reporting the bug you encountered.

### Bug Reproduction

The bug you describe is not sufficiently well described for it to be reproduced on my machine. I invite you to share the lines of code that led you to the error you encountered (in particular by replacing the template in the "To Reproduce / Steps to reproduce the behavior" step) and to share the error you encountered in the console. 

As I'll describe in the next section, I think the error is due to the size of your calibration set. But I don't have enough information to confirm this 100%. I thank you in advance for your help.

### Bug Description

TL;DR: A priori, this is not a problem with `prefit` mode. This is a problem that can arise when the calibration set and test test have a larger number of samples. This behavior is unintended, as the `predict` method is generally used with a smaller number of test samples during inference.

With the elements you've shared with me, I think your problem is close to that of another issue #326 I encountered a short while ago (see this link: https://github.com/scikit-learn-contrib/MAPIE/issues/326#issuecomment-1632337436).

Concerning the behaviour of the `predict` method, as explained in the [theoretical description of the documentation](https://mapie.readthedocs.io/en/latest/theoretical_description_regression.html) (see Fig of CV+), MAPIE needs to compute the distribution of residuals and predictions for all calibration samples for each test point. MAPIE does it in a vectorized way with two-dimensional arrays of size `(n_calib_samples, n_test_samples)`, hence exceeding the memory when the number of calibration and test samples are high.

In your case, your calibration set and test set are the same and therefore of the same size. So, during prediction, you generate a matrix of size `(n_calib_samples, n_calib_samples)`, which if too large, exhausts all RAM.

### Conclusion

I hope these answers have helped you. Don't hesitate to share your lines of code with me to reproduce your error and better understand your bug.

**Comment by artem-math** [comment]
Hello @thibaultcordier ! You are absolutely right, the issue was with the number of samples I have for calibration and prediction. My dataset is around 2 500 000 observations, so any typical split for train/test/validate would cause this pain =( I will consider the workaround that you have proposed here #326 (comment)

Thank you =)
